storage/testonly/admin_storage_tester.go

```diff
@@ -331,188 +331,276 @@ func (tester *AdminStorageTester) TestUpdateTree(t *testing.T) {
 		desc string
-		// newTrees is the number of trees created before each test. The total of trees is
-		// cumulative between tests.
-		newTrees int
+		tree *trillian.Tree
 	}{
-		{desc: "zero", newTrees: 0},
-		{desc: "one", newTrees: 1},  // total 1
-		{desc: "four", newTrees: 3}, // total 4
+		{desc: "logTree", tree: logTree},
+		{desc: "mapTree", tree: mapTree},
 	}
+	for _, test := range tests {
+		deletedTree, err := softDeleteTree(ctx, s, test.tree.TreeId)
+		if err != nil {
+			t.Errorf("%v: softDeleteTree() returned err = %v", test.desc, err)
+			continue
+		}
+
+		if deletedTree.GetDeleteTime().GetSeconds() == 0 {
+			t.Errorf("%v: tree.DeleteTime = %v, want > 0", test.desc, deletedTree.DeleteTime)
+		}
+
+		wantTree := proto.Clone(test.tree).(*trillian.Tree)
+		wantTree.Deleted = true
+		wantTree.DeleteTime = deletedTree.DeleteTime
+		if got, want := deletedTree, wantTree; !proto.Equal(got, want) {
+			t.Errorf("%v: post-softDeleteTree diff (-got +want):\n%v", test.desc, pretty.Compare(got, want))
+		}
+
+		if err := assertStoredTree(ctx, s, deletedTree); err != nil {
+			t.Errorf("%v: %v", test.desc, err)
+		}
+	}
+}
 
+// TestSoftDeleteTreeErrors tests error scenarios of SoftDeleteTree.
+func (tester *AdminStorageTester) TestSoftDeleteTreeErrors(t *testing.T) {
 	ctx := context.Background()
 	s := tester.NewAdminStorage()
 
-	wantTrees := []*trillian.Tree{}
-	for _, test := range tests {
-		func() {
-			// Setup new trees (as specified)
-			if test.newTrees > 0 {
-				before := len(wantTrees)
-				for j := 0; j < test.newTrees; j++ {
-					tree, err := createTree(ctx, s, LogTree)
-					if err != nil {
-						t.Fatalf("%v: CreateTree() = (_, %v), want = (_, nil)", test.desc, err)
-					}
-					wantTrees = append(wantTrees, tree)
-				}
-				if got := len(wantTrees) - before; got != test.newTrees {
-					t.Fatalf("%v: got %v new trees, want = %v", test.desc, got, test.newTrees)
-				}
-			}
+	softDeleted := makeTreeOrFail(ctx, t.Fatalf, s, spec{Tree: LogTree, Deleted: true})
 
-			tx, err := s.Snapshot(ctx)
-			if err != nil {
-				t.Fatalf("%v: Snapshot() = %v, want = nil", test.desc, err)
-			}
-			defer tx.Close()
-			if err := runListTreeIDsTest(ctx, tx, wantTrees); err != nil {
-				t.Errorf("%v: %v", test.desc, err)
-			}
-			if err := runListTreesTest(ctx, tx, wantTrees); err != nil {
-				t.Errorf("%v: %v", test.desc, err)
-			}
-			if err := tx.Commit(); err != nil {
-				t.Errorf("%v: Commit() = %v, want = nil", test.desc, err)
-			}
-		}()
+	tests := []struct {
+		desc     string
+		treeID   int64
+		wantCode errors.Code
+	}{
+		{desc: "unknownTree", treeID: 12345, wantCode: errors.NotFound},
+		{desc: "alreadyDeleted", treeID: softDeleted.TreeId, wantCode: errors.FailedPrecondition},
+	}
+	for _, test := range tests {
+		if _, err := softDeleteTree(ctx, s, test.treeID); errors.ErrorCode(err) != test.wantCode {
+			t.Errorf("%v: softDeleteTree() returned err = %v, wantCode = %s", test.desc, err, test.wantCode)
+		}
 	}
 }
 
```

Comment:
I wonder whether it's worth having a `TestFullDelete` case that does the sequence that would occur in practice: create, (maybe) freeze, soft-delete, hard-delete.  

(However, all the individual pieces are already present and checked, so it wouldn't test anything new.)

Reply:
That's what happens on "TestHardDeleteTree" (apart from the freeze step). I'll add a frozen tree to the test.

Reply:
Ah yes, I guess it does, hidden underneath the `makeTreeOrFail`  -- fine.